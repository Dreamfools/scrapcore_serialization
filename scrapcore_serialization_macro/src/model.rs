use crate::error::{bail, MacroError};
use crate::model::attrs::{ModelAttributeConfig, SharedAttributeConfig};
use crate::model::enums::{process_enum, EnumVariantAttributeInput};
use crate::model::structs::{process_struct, FieldAttributeInput};
use crate::{serialized_of, MOD_ERRORS, MOD_REGISTRY, MOD_SERIALIZATION};
use darling::ast::Data;
use darling::FromDeriveInput;
use proc_macro2::{Ident, TokenStream};
use quote::{quote, quote_spanned};
use std::borrow::Cow;

use darling::util::SpannedValue;
use std::ops::Deref;
use syn::spanned::Spanned;
use syn::{DeriveInput, Meta, Type, WhereClause};

mod enums;
mod structs;

mod attrs;

/// Derive attribute
#[derive(Debug, FromDeriveInput)]
#[darling(
    attributes(my_crate),
    forward_attrs(model_attr, model_serde),
    supports(struct_named, enum_newtype, enum_unit)
)]
pub struct ModelAttributeInput {
    // Internal fields
    ident: Ident,

    generics: syn::Generics,

    attrs: Vec<syn::Attribute>,

    // Configurable fields
    #[darling(flatten)]
    config: ModelAttributeConfig,

    data: Data<SpannedValue<EnumVariantAttributeInput>, SpannedValue<FieldAttributeInput>>,
}

impl ModelAttributeConfig {
    fn schema_derive(&self) -> Option<TokenStream> {
        if self.no_schema.is_present() {
            None
        } else {
            Some(quote! {
                #[derive(schemars::JsonSchema)]
            })
        }
    }

    fn where_clauses(&self) -> impl Iterator<Item = TokenStream> + '_ {
        self.extra_conditions.iter().map(|clause| {
            let predicates = &clause.predicates;
            quote!(#predicates)
        })
    }
}

pub fn model_macro_impl(input: DeriveInput) -> Result<TokenStream, MacroError> {
    let data = ModelAttributeInput::from_derive_input(&input)?;

    match data.data {
        Data::Enum(variants) => {
            process_enum(data.ident, data.generics, data.attrs, data.config, variants)
        }
        Data::Struct(fields) => {
            process_struct(data.ident, data.generics, data.attrs, data.config, fields)
        }
    }
}

fn fallthrough(attrs: &[syn::Attribute]) -> Vec<TokenStream> {
    let mut inner_attrs = vec![];
    for attr in attrs {
        let Meta::List(list) = &attr.meta else {
            continue;
        };
        let Some(name) = list.path.segments.last() else {
            continue;
        };
        let tokens = &list.tokens;
        match name.ident.to_string().as_str() {
            "model_attr" => {
                inner_attrs.push(quote_spanned!(attr.span()=>#[#tokens]));
            }
            "model_serde" => {
                inner_attrs.push(quote_spanned!(attr.span()=>#[serde(#tokens)]));
            }
            _ => continue,
        }
    }
    inner_attrs
}

fn edit_where_clause(
    clause: Option<&WhereClause>,
    fields_conditions: impl IntoIterator<Item = TokenStream>,
    extra_conditions: impl IntoIterator<Item = TokenStream>,
) -> TokenStream {
    let reg = MOD_REGISTRY.deref();
    let fields_conditions = fields_conditions.into_iter();
    let extra_conditions = extra_conditions.into_iter();
    if let Some(clause) = clause {
        quote! {
            #clause, #(#fields_conditions,)* #(#extra_conditions,)* Registry: #reg::PartialRegistry
        }
    } else {
        quote! {
            where #(#fields_conditions,)* #(#extra_conditions,)* Registry: #reg::PartialRegistry
        }
    }
}

impl SharedAttributeConfig {
    fn serialized_ty<'a>(&'a self, field_ty: &'a Type) -> Result<Cow<'a, Type>, MacroError> {
        // "raw" fields use their target type directly, otherwise lookup
        // `#[model(ty=T)]` or use `SerializationFallback`
        let field_ty = self.from.as_ref().unwrap_or(field_ty);
        let ty = if self.raw.is_present() {
            if self.custom_ty.is_some() || self.with.is_some() {
                bail!(self.raw.span(), "`raw` attribute field can not be used at the same time as `ty` or `with` attributes")
            }
            Cow::Borrowed(field_ty)
        } else if let Some(ty) = &self.custom_ty {
            Cow::Borrowed(ty)
        } else {
            Cow::Owned(serialized_of(field_ty)?)
        };

        Ok(ty)
    }

    fn where_condition(&self, original_type: &Type, serialized_type: &Type) -> Option<TokenStream> {
        let ser = MOD_SERIALIZATION.deref();
        let original_type = self.from.as_ref().unwrap_or(original_type);
        (self.with.is_none() && !self.raw.is_present()).then(|| {
            quote! {
                #serialized_type: #ser::DeserializeModel::<#original_type, Registry>
            }
        })
    }

    fn deserialization_code(
        &self,
        field_ty: &Type,
        varname: &Ident,
        context: Option<TokenStream>,
    ) -> Result<TokenStream, MacroError> {
        let ser = MOD_SERIALIZATION.deref();
        let err = MOD_ERRORS.deref();

        let target_type = self.from.as_ref().unwrap_or(field_ty);
        let mut blocks = vec![];
        if self.raw.is_present() {
            blocks.push(quote! {
                let #varname: #target_type = #varname;
            })
        } else if let Some(func) = &self.with {
            blocks.push(quote! {
                let #varname: #target_type = #func(#varname, registry)?;
            })
        } else {
            blocks.push(quote! {
                let #varname: #target_type = #ser::DeserializeModel::<#target_type, Registry>::deserialize(#varname, registry)?;
            })
        };

        if let Some(min) = &self.min {
            blocks.push(quote! {
                let #varname: #target_type = #ser::ApplyMin::apply(#varname, #min)?;
            })
        }

        if let Some(max) = &self.max {
            blocks.push(quote! {
                let #varname: #target_type = #ser::ApplyMax::apply(#varname, #max)?;
            })
        }

        if let Some(from) = &self.from {
            blocks.push(quote! {
                let #varname: #field_ty = <#field_ty as From<#from>>::from(#varname);
            });
        } else {
            assert_eq!(field_ty, target_type)
        }

        let context = match context {
            None => quote!(None),
            Some(ctx) => {
                quote! {
                    Some(||#ctx)
                }
            }
        };

        let code = quote! {
            #err::s_try(&mut *registry, |registry: &mut Registry| {
                #(#blocks)*
                Ok(#varname)
            }, #context)?
        };

        Ok(code)
    }
}
