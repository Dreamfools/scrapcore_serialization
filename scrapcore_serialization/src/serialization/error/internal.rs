use crate::registry::path_identifier::PathIdentifier;
use crate::registry::SerializationRegistry;
use crate::serialization::error::{DeserializationError, DeserializationErrorKind};
use crate::ItemId;
use thiserror::Error;

/// Internal errors that are results of bad application code organization,
/// rather than loaded mod issues
#[derive(Debug, Error, Clone)]
pub enum InternalDeserializationError<Registry: SerializationRegistry> {
    #[error(
        "Partial registry is poisoned. This is likely caused by ignoring some previous errors."
    )]
    PoisonedRegistry,
    #[error("Collection item was not registered before deserializing")]
    EntryNotRegistered,
    #[error("Reserved collection entry was deleted during item deserialization")]
    EntryGoneDuringDeserialization,
    #[error("Reserved {} collection entry was turned into Raw entry during deserialization: {}", .1, .0)]
    EntryBecameRaw(PathIdentifier, Registry::ItemKind),
    #[error("Reserved {} collection entry was turned into a different Deserialized entry during deserialization: {}", .1, .0)]
    EntryBecameDeserialized(PathIdentifier, Registry::ItemKind),
    #[error("Reserved {} collection entry has different ID after deserialization: {}", .1, .0,)]
    EntryChangedId(ItemId, Registry::ItemKind),
    #[error("Attempted to deserialize {} collection item while having a conflicting raw entry: {}", .1, .0)]
    ConflictingRawEntry(PathIdentifier, Registry::ItemKind),
    #[error("Attempted to deserialize {} collection item while having a conflicting deserialized entry: {}", .1, .0)]
    ConflictingDeserializedEntry(PathIdentifier, Registry::ItemKind),
    #[error("{} collection item was not deserialized before conversion: {}, was `process_raw_singleton` not invoked?", .1, .0)]
    ConversionEntryNotDeserialized(ItemId, Registry::ItemKind),
    #[error("{} collection item was left in reserved state before conversion: {}, was registry poisoned?", .1, .0)]
    ConversionEntryReserved(ItemId, Registry::ItemKind),
    #[error("Got divergence in collection IDs for item {}({}): expected to be saved at ID {} but instead got {}", .kind, .key, .expected, .got)]
    ConversionIdsDiverge {
        key: ItemId,
        expected: usize,
        got: usize,
        kind: Registry::ItemKind,
    },
    #[error("Singleton {} is missing during conversion, this should have been caught during `process_raw_singleton` invocation", .0)]
    ConversionMissingSingleton(Registry::ItemKind),
    #[error("Singleton {}({}) is not processed, was `process_raw_singleton` not invoked?", .1, .0)]
    ConversionUnprocessedSingleton(PathIdentifier, Registry::ItemKind),
}

impl<Registry: SerializationRegistry> InternalDeserializationError<Registry> {
    pub fn into_err(self) -> DeserializationError<Registry> {
        self.into()
    }
}

impl<Registry: SerializationRegistry> From<InternalDeserializationError<Registry>>
    for DeserializationError<Registry>
{
    fn from(value: InternalDeserializationError<Registry>) -> Self {
        DeserializationErrorKind::InternalError(value).into()
    }
}
