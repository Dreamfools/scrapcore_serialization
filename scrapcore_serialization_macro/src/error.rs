use proc_macro2::TokenStream;
use thiserror::Error;

pub fn tokens(result: Result<TokenStream, MacroError>) -> TokenStream {
    result.unwrap_or_else(|err| err.into_tokens())
}

#[derive(Debug, Error)]
#[error("{}", .0)]
pub enum MacroError {
    Syn(syn::Error),
    Darling(darling::Error),
}

impl MacroError {
    pub fn into_tokens(self) -> TokenStream {
        match self {
            MacroError::Syn(err) => err.to_compile_error(),
            MacroError::Darling(err) => err.write_errors(),
        }
    }
}

impl From<syn::Error> for MacroError {
    fn from(err: syn::Error) -> Self {
        MacroError::Syn(err)
    }
}

impl From<darling::Error> for MacroError {
    fn from(err: darling::Error) -> Self {
        MacroError::Darling(err)
    }
}

macro_rules! error {
    ($span:expr, $($t:tt)*) => {
        syn::Error::new($span, format!($($t)*))
    };
}

pub(crate) use error;

macro_rules! bail {
    ($span:expr, $($t:tt)*) => {
        return Err($crate::error::error!($span, $($t)*).into())
    };
}

pub(crate) use bail;
