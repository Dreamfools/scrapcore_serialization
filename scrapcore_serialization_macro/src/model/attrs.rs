use darling::util::Flag;
use darling::FromMeta;
use proc_macro2::Ident;
use syn::{Expr, Path, Type};

/// Attribute fields only usable on the struct/enum-level #\[model] attribute
#[derive(Debug, FromMeta)]
pub struct ModelAttributeConfig {
    /// Custom name for the serialized type
    pub name: Option<Ident>,
    /// Skips deriving `schemars` schema
    pub no_schema: Flag,
    /// Extra "where" conditions for the `DeserializeModel` implementation
    #[darling(multiple)]
    #[darling(rename = "condition")]
    pub extra_conditions: Vec<syn::WhereClause>,
}

/// Attribute fields for both struct and enum fields
#[derive(Debug, FromMeta)]
pub struct SharedAttributeConfig {
    /// Custom serialized field type
    ///
    /// This flag is **conflicting** with `raw` flag
    #[darling(rename = "ty")]
    pub custom_ty: Option<Type>,
    /// Path to the custom deserialization function for the field
    ///
    /// Deserialization function must have signature
    /// `fn(S, Registry) -> Result<T, impl DeserializationError<Registry>>`,
    /// where `S` is a serialized field type (or type specified in `ty`
    /// attribute), and `T` is the resulting field type (or type specified in
    /// `from` attribute)
    ///
    /// This flag is **conflicting** with `raw` flag
    pub with: Option<Path>,
    /// Completely skips deserialization for the given field
    ///
    /// This flag is identical to using`#[model(ty=T, with=identity_function)]`,
    /// where `T` is the field type, and so this flag is **conflicting** with
    /// `ty` and `with` flags
    ///
    /// Useful for types from external crates that do not implement `DeserializeModel`
    pub raw: Flag,

    /// Forces the field to be deserialized as a given type,
    /// and then converted into the field type using the `From` impl
    pub from: Option<Type>,

    /// Applies min validator to the field
    pub min: Option<Expr>,
    /// Applies max validator to the field
    pub max: Option<Expr>,

    /// Marks the field as "Id" field, emitting a different `where` condition.
    ///
    /// Often useful for avoiding recursive trait bounds.
    ///
    /// Relies on `ReverseId` trait:
    /// `where Registry: PartialCollectionHolder<<T as ReverseId>::Item>`
    /// where `T` is the field type
    pub id: Flag,

    /// Same as [id], but uses the provided type instead of relying on a trait
    ///
    /// Often useful for avoiding recursive trait bounds.
    ///
    /// Example generated condition:
    /// `where Registry: PartialCollectionHolder<T>` where `T` is the provided type
    pub id_of: Option<Type>,

    /// Skips generating `where` condition for this field
    ///
    /// Useful for avoiding recursion issues
    pub no_condition: Flag,
}

/// Attribute fields for struct fields
#[derive(Debug, FromMeta)]
pub struct StructFieldAttributeConfig {
    /// Generated AsRef implementation for marked struct to value of this field
    pub as_ref: Flag,
    /// Custom name for the serialized field
    pub rename: Option<Ident>,

    /// Shared config, see [SharedAttributeConfig] for available attributes
    #[darling(flatten)]
    pub config: SharedAttributeConfig,
}

/// Attribute fields for enum fields
#[derive(Debug, FromMeta)]
pub struct EnumVariantAttributeConfig {
    /// Custom name for the serialized variant
    pub rename: Option<Ident>,
    /// Shared config, see [SharedAttributeConfig] for available attributes
    #[darling(flatten)]
    pub config: SharedAttributeConfig,
}
