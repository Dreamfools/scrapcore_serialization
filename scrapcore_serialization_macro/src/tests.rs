use crate::error::tokens;
use crate::model::model_macro_impl;

use proc_macro2::TokenStream;
use quote::TokenStreamExt;

use std::path::Path;

fn check_model(path: &Path) -> String {
    let data = std::fs::read_to_string(path).expect("Should be able to read the file");
    let mut stream: TokenStream = data
        .parse()
        .expect("Test case data should be a valid rust file");
    let tokens = tokens(model_macro_impl(
        syn::parse2(stream.clone()).expect("Test input should be a valid derive input"),
    ));

    stream.append_all(tokens);

    let file: syn::File = syn::parse2(stream).expect("Should generate valid rust code");

    prettyplease::unparse(&file)
}

#[test]
fn struct_tests() {
    insta::glob!("test_cases/model/struct/*.rs", |path| {
        insta::assert_snapshot!(check_model(path))
    });
}
