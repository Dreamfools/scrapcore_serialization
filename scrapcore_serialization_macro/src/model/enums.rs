use crate::error::{bail, MacroError};
use crate::model::attrs::{EnumVariantAttributeConfig, ModelAttributeConfig};
use crate::model::{edit_where_clause, fallthrough};
use crate::{MOD_ERRORS, MOD_REGISTRY, MOD_SERIALIZATION};
use darling::ast::Fields;
use darling::util::SpannedValue;
use darling::{FromField, FromMeta, FromVariant};
use itertools::Itertools;
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote, quote_spanned};
use std::ops::Deref;
use syn::{GenericParam, TypeParam};

#[derive(Debug, FromVariant)]
#[darling(attributes(model))]
pub struct EnumVariantAttributeInput {
    ident: Ident,
    fields: Fields<EnumVariantField>,
    attrs: Vec<syn::Attribute>,

    #[darling(flatten)]
    config: EnumVariantAttributeConfig,
}

#[derive(Debug, FromField)]
struct EnumVariantField {
    ident: Option<Ident>,
    ty: syn::Type,
}

pub fn process_enum(
    model_name: Ident,
    mut generics: syn::Generics,
    attributes: Vec<syn::Attribute>,
    config: ModelAttributeConfig,
    variants: Vec<SpannedValue<EnumVariantAttributeInput>>,
) -> Result<TokenStream, MacroError> {
    let model_fallthrough_attrs = fallthrough(&attributes);
    let serialized_name = config
        .name
        .as_ref()
        .map(|e| format_ident!("{}", e))
        .unwrap_or_else(|| format_ident!("{}Serialized", model_name));

    let _reg = MOD_REGISTRY.deref();
    let ser = MOD_SERIALIZATION.deref();
    let err = MOD_ERRORS.deref();

    let variants = variants.iter().map(|variant| {
        let field = match variant.fields.iter().at_most_one() {
            Ok(field) => field,
            Err(_) => {
                bail!(
                    variant.span(),
                    "Enum variant must have exactly zero or one field"
                );
            }
        };

        let fallthrough_attrs = fallthrough(&variant.attrs);
        let variant_name = variant.config.rename.as_ref().unwrap_or(&variant.ident);

        let (serialized_variant, deserialization_match, where_condition) =
            if let Some(field) = field {
                if field.ident.is_some() {
                    bail!(variant.span(), "Only newtype enums are supported");
                }

                let serialized_ty = variant.config.config.serialized_ty(&field.ty)?;

                let serialized_variant = quote_spanned! {variant.span()=>
                    #(#fallthrough_attrs)*
                    #variant_name(#serialized_ty),
                };

                let item_var = Ident::new("item", variant.span());

                let deser_code = variant.config.config.deserialization_code(
                &field.ty,
                &item_var,
                Some(
                    quote!(#err::DeserializationErrorStackItem::Variant(stringify!(#variant_name))),
                ),
            )?;

                let deserialization_match = quote_spanned! {variant.span()=>
                    Self::#variant_name(#item_var) => #model_name::#variant_name(#deser_code),
                };

                let where_condition = variant
                    .config
                    .config
                    .where_condition(&field.ty, &serialized_ty)?;

                (serialized_variant, deserialization_match, where_condition)
            } else {
                let serialized_variant = quote_spanned! {variant.span()=>
                    #(#fallthrough_attrs)*
                    #variant_name,
                };
                let deserialization_match = quote_spanned! {variant.span()=>
                    Self::#variant_name => #model_name::#variant_name,
                };
                let where_condition = None;
                (serialized_variant, deserialization_match, where_condition)
            };

        Result::<(TokenStream, TokenStream, Option<TokenStream>), MacroError>::Ok((
            serialized_variant,
            deserialization_match,
            where_condition,
        ))
    });

    let (members, deser, where_conditions) = itertools::process_results(variants, |iter| {
        iter.multiunzip::<(Vec<_>, Vec<_>, Vec<_>)>()
    })?;

    let where_conditions = where_conditions.into_iter().flatten();

    let schema_derive = config.schema_derive();
    let (gen_imp, gen_ty, gen_wher) = generics.split_for_impl();
    let defs = quote! {
        #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
        #schema_derive
        #(#model_fallthrough_attrs)*
        pub enum #serialized_name {
            #(#members)*
        }

        #[automatically_derived]
        impl #gen_imp #ser::SerializationFallback for #model_name #gen_ty #gen_wher {
            type Fallback = #serialized_name;
        }

        #[automatically_derived]
        impl #gen_imp  AsRef<#model_name #gen_ty> for #model_name #gen_ty #gen_wher {
            fn as_ref(&self) -> &#model_name {
                &self
            }
        }

        // #[automatically_derived]
        // impl #serialization_mod::ModelDeserializable<#model_name> for #serialized_name {
        //     fn deserialize(self, registry: &mut #model_mod::PartialModRegistry) -> Result<#model_name, #serialization_mod::DeserializationError> {
        //         Ok(match self {
        //             #(#deserialization)*
        //         })
        //     }
        // }
    };

    let gen_ty = quote!(#gen_ty);

    generics
        .params
        .push(GenericParam::Type(TypeParam::from_string("Registry")?));

    let (gen_imp, _, gen_wher) = generics.split_for_impl();

    let where_condition = edit_where_clause(gen_wher, where_conditions, config.where_clauses());

    let deser_code = quote! {
        #[automatically_derived]
        impl #gen_imp #ser::DeserializeModel<#model_name, Registry> for #serialized_name #gen_ty #where_condition {
            fn deserialize(self, registry: &mut Registry) -> Result<#model_name, #err::DeserializationError<Registry>> {
                Ok(match self {
                    #(#deser)*
                })
            }
        }
    };

    Ok(quote! {
        #defs
        #deser_code
    })
}
