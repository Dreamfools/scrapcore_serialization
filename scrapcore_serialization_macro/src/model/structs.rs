use crate::error::{bail, MacroError};
use crate::model::attrs::{
    ModelAttributeConfig, SharedAttributeConfig, StructFieldAttributeConfig,
};
use crate::model::{edit_where_clause, fallthrough};
use crate::{MOD_ERRORS, MOD_REGISTRY, MOD_SERIALIZATION};
use convert_case::{Case, Casing};
use darling::ast::Fields;
use darling::util::SpannedValue;
use darling::{FromField, FromMeta};
use itertools::Itertools;
use proc_macro2::TokenStream;
use quote::{format_ident, quote, quote_spanned};
use std::borrow::Cow;
use std::ops::Deref;
use syn::spanned::Spanned;
use syn::{GenericParam, Ident, Type, TypeParam};

#[derive(Debug, FromField)]
#[darling(attributes(model), forward_attrs(model_attr, model_serde))]
pub struct FieldAttributeInput {
    ident: Option<syn::Ident>,
    ty: Type,
    attrs: Vec<syn::Attribute>,
    #[darling(flatten)]
    config: StructFieldAttributeConfig,
}

#[derive(Debug)]
struct FieldData<'a> {
    name: Ident,
    original_type: &'a Type,
    serialized_type: Cow<'a, Type>,
    definition: TokenStream,
    config: &'a SharedAttributeConfig,
}

pub fn process_struct(
    model_name: Ident,
    mut generics: syn::Generics,
    attributes: Vec<syn::Attribute>,
    config: ModelAttributeConfig,
    struct_fields: Fields<SpannedValue<FieldAttributeInput>>,
) -> Result<TokenStream, MacroError> {
    let mut fields = Vec::new();
    let model_fallthrough_attrs = fallthrough(&attributes);

    let serialized_name = config
        .name
        .as_ref()
        .map(|e| format_ident!("{}", e))
        .unwrap_or_else(|| format_ident!("{}Serialized", model_name));

    let mut as_refs = vec![];

    for field in &struct_fields.fields {
        let Some(name) = &field.ident else {
            bail!(field.ty.span(), "All model fields must be named")
        };
        let ty = &field.ty;

        if field.config.as_ref.is_present() {
            as_refs.push(quote_spanned! {name.span()=>
                #[automatically_derived]
                impl AsRef<#ty> for #model_name {
                    fn as_ref(&self) -> &#ty {
                        &self.#name
                    }
                }
            })
        }

        let serialized_type = field.config.config.serialized_ty(ty)?;
        let fallthrough_attrs = fallthrough(&field.attrs);
        let definition = quote_spanned!(name.span()=>
            #(#fallthrough_attrs)*
            #name: #serialized_type
        );

        let field_data = FieldData {
            name: name.clone(),
            definition,
            original_type: ty,
            serialized_type,
            config: &field.config.config,
        };

        fields.push(field_data)
    }

    let tokens = fields.iter().map(|e| &e.definition);
    let schema_derive = config.schema_derive();

    let _reg = MOD_REGISTRY.deref();
    let ser = MOD_SERIALIZATION.deref();
    let err = MOD_ERRORS.deref();

    let (gen_imp, gen_ty, gen_wher) = generics.split_for_impl();

    let serialized_struct = quote!(
        #(#model_fallthrough_attrs)*
        #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
        #schema_derive
        #[serde(rename_all = "camelCase")]
        pub struct #serialized_name #generics {
            #(#tokens),*
        }

        #[automatically_derived]
        impl #gen_imp #ser::SerializationFallback for #model_name #gen_ty #gen_wher {
            type Fallback = #serialized_name;
        }

        #(#as_refs)*

        #[automatically_derived]
        impl #gen_imp  AsRef<#model_name #gen_ty> for #model_name #gen_ty #gen_wher {
            fn as_ref(&self) -> &#model_name {
                &self
            }
        }
    );

    let map_name = config.name.clone().unwrap_or_else(|| {
        format_ident!(
            "{}",
            model_name
                .to_string()
                .from_case(Case::Pascal)
                .to_case(Case::Snake)
        )
    });
    let _kind_name = format_ident!(
        "{}",
        map_name
            .to_string()
            .from_case(Case::Snake)
            .to_case(Case::Pascal)
    );
    // let _map_name = format_ident!("{}", map_name);
    // let _kind_name = format_ident!("{}", kind_name);

    let _names = fields.iter().map(|e| &e.name);

    let serialized_object_name = format_ident!("serialized");

    let modifiers: Vec<TokenStream> = fields
        .iter()
        .map(|f| {
            let name = &f.name;
            let data = &Ident::new("item", name.span());
            let original_type = &f.original_type;
            let deser_code = f.config.deserialization_code(
                original_type,
                data,
                Some(quote!(#err::DeserializationErrorStackItem::Field(stringify!(#name)))),
            )?;
            Result::<TokenStream, MacroError>::Ok(quote_spanned! { original_type.span()=>
                #name: {
                    let #data = #serialized_object_name.#name;
                    #deser_code
                },
            })
        })
        .try_collect()?;

    let field_where_conditions = fields.iter().map(
        |FieldData {
             original_type,
             serialized_type,
             config,
             ..
         }| { config.where_condition(original_type, serialized_type) },
    );

    let field_where_conditions =
        itertools::process_results(field_where_conditions, |i| i.flatten().collect_vec())?;

    let (_, gen_ty, _) = generics.split_for_impl();
    let gen_ty = quote!(#gen_ty);

    generics
        .params
        .push(GenericParam::Type(TypeParam::from_string("Registry")?));

    let (gen_imp, _, gen_wher) = generics.split_for_impl();

    let where_condition =
        edit_where_clause(gen_wher, field_where_conditions, config.where_clauses());

    let deserialization_impl = quote! {
        #[automatically_derived]
        impl #gen_imp #ser::DeserializeModel<#model_name, Registry> for #serialized_name #gen_ty #where_condition {
            fn deserialize(self, registry: &mut Registry) -> Result<#model_name, #err::DeserializationError<Registry>> {
                let #serialized_object_name = self;

                Ok(#model_name {
                    #(#modifiers)*
                })
            }
        }
    };

    let all_together = quote! {
        #serialized_struct

        #deserialization_impl
    };

    Ok(all_together)
}
