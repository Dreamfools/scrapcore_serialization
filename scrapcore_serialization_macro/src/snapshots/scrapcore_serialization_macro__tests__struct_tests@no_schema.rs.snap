---
source: scrapcore_serialization_macro/src/tests.rs
expression: check_model(path)
input_file: scrapcore_serialization_macro/src/test_cases/model/struct/no_schema.rs
---
#[derive(Debug, DatabaseModel)]
#[model(no_schema)]
struct Test {
    a: u32,
    b: String,
    c: Vec<Test>,
}
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TestSerialized {
    a: <u32 as scrapcore_serialization::serialization::SerializationFallback>::Fallback,
    b: <String as scrapcore_serialization::serialization::SerializationFallback>::Fallback,
    c: <Vec<
        Test,
    > as scrapcore_serialization::serialization::SerializationFallback>::Fallback,
}
#[automatically_derived]
impl scrapcore_serialization::serialization::SerializationFallback for Test {
    type Fallback = TestSerialized;
}
#[automatically_derived]
impl AsRef<Test> for Test {
    fn as_ref(&self) -> &Test {
        &self
    }
}
#[automatically_derived]
impl<Registry> scrapcore_serialization::serialization::DeserializeModel<Test, Registry>
for TestSerialized
where
    <u32 as scrapcore_serialization::serialization::SerializationFallback>::Fallback: scrapcore_serialization::serialization::DeserializeModel<
        u32,
        Registry,
    >,
    <String as scrapcore_serialization::serialization::SerializationFallback>::Fallback: scrapcore_serialization::serialization::DeserializeModel<
        String,
        Registry,
    >,
    <Vec<
        Test,
    > as scrapcore_serialization::serialization::SerializationFallback>::Fallback: scrapcore_serialization::serialization::DeserializeModel<
        Vec<Test>,
        Registry,
    >,
    Registry: scrapcore_serialization::registry::PartialRegistry,
{
    fn deserialize(
        self,
        registry: &mut Registry,
    ) -> Result<
        Test,
        scrapcore_serialization::serialization::error::DeserializationError<Registry>,
    > {
        let serialized = self;
        Ok(Test {
            a: {
                let item = serialized.a;
                scrapcore_serialization::serialization::error::s_try(
                    &mut *registry,
                    |registry: &mut Registry| {
                        let item: u32 = scrapcore_serialization::serialization::DeserializeModel::<
                            u32,
                            Registry,
                        >::deserialize(item, registry)?;
                        Ok(item)
                    },
                    Some(|| scrapcore_serialization::serialization::error::DeserializationErrorStackItem::Field(
                        stringify!(a),
                    )),
                )?
            },
            b: {
                let item = serialized.b;
                scrapcore_serialization::serialization::error::s_try(
                    &mut *registry,
                    |registry: &mut Registry| {
                        let item: String = scrapcore_serialization::serialization::DeserializeModel::<
                            String,
                            Registry,
                        >::deserialize(item, registry)?;
                        Ok(item)
                    },
                    Some(|| scrapcore_serialization::serialization::error::DeserializationErrorStackItem::Field(
                        stringify!(b),
                    )),
                )?
            },
            c: {
                let item = serialized.c;
                scrapcore_serialization::serialization::error::s_try(
                    &mut *registry,
                    |registry: &mut Registry| {
                        let item: Vec<Test> = scrapcore_serialization::serialization::DeserializeModel::<
                            Vec<Test>,
                            Registry,
                        >::deserialize(item, registry)?;
                        Ok(item)
                    },
                    Some(|| scrapcore_serialization::serialization::error::DeserializationErrorStackItem::Field(
                        stringify!(c),
                    )),
                )?
            },
        })
    }
}
